package/base-files/files/lib/functions/system.sh

```diff
@@ -86,6 +86,23 @@ mtd_get_mac_ascii() {
 	get_mac_ascii "$part" "$key"
 }
 
+mtd_get_mac_ascii_ubi() {
+	local mtdname="$1"
+	local ubivolname="$2"
+	local key="$3"
+	local part
+	local mac_dirty
+
+	. /lib/upgrade/nand.sh
+
+	ubidev="$(nand_find_ubi "$mtdname")"
+	[ -n "$ubidev" ] && ubivol="$(nand_find_volume "$ubidev" "$ubivolname")"
+	[ -n "$ubivol" ] && mac_dirty=$(strings "/dev/$ubivol" | sed -n 's/^'"$key"'=//p')
+
+	# "canonicalize" mac
+	[ -n "$mac_dirty" ] && macaddr_canonicalize "$mac_dirty"
+}
```

Comment:
Please make use of `get_mac_ascii` like `mmc_get_mac_ascii` does

Reply:
Actually we got support for NVMEM-on-UBI by now and you may use that and not mess around with MAC addresses in userspace to begin with.

See device tree of ASUS-TUX4200 or ASUS-TUX6000, for example.

Reply:
> Actually we got support for NVMEM-on-UBI by now and you may use that and not mess around with MAC addresses in userspace to begin with.
> 
> See device tree of ASUS-TUX4200 or ASUS-TUX6000, for example.

@dangowrt I tried like this below, but it does not appear to work. I would need an overlay over u-boot env, from which I need to fetch 'ethaddr' in ascii.

```
partition@100000 {
	reg = <0x100000 0xff00000>;
	label = "ubi";
	compatible = "linux,ubi";

	volumes {
		ubi-volume-env1 {
			volname = "env1";
			nvmem-layout {
				compatible = "u-boot,env-redundant-bool";

				ethaddr: ethaddr {
					#nvmem-cell-cells = <1>;
				};
			};

		};

		ubi-volume-env2 {
			volname = "env2";
			nvmem-layout {
				compatible = "u-boot,env-redundant-bool";
			};
		};
	};
};

&gmac0 {
	nvmem-cells = <&ethaddr>;
	nvmem-cell-names = "mac-address";
};

```

Reply:
Forgot that ramips is still on Linux 5.15... Once we dropped 5.15 you will also need to also enable `CONFIG_MTD_UBI_NVMEM=y` in `target/linux/ramips/mt7621/config-6.1` for this to work...